clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+static std::string constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
+                                       MangleContext &MC) {
+  QualType KernelNameType =
+      calculateKernelNameType(S.getASTContext(), KernelCallerFunc);
+
+  if (S.getLangOpts().SYCLUnnamedLambda)
+    return PredefinedExpr::ComputeName(
+        S.getASTContext(), PredefinedExpr::UniqueStableNameType, KernelNameType);
+
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+
+  MC.mangleTypeName(KernelNameType, Out);
+  return std::string(Out.str());
+}
+
+// anonymous namespace so these don't get linkage.
+namespace {
+// A base type that the SYCL OpenCL Kernel construction task uses to implement
+// individual tasks.
+template<typename Derived>
+class SyclKernelFieldHandler {
+protected:
+  Sema &SemaRef;
+  SyclKernelFieldHandler(Sema &S) : SemaRef(S) {}
+
+public:
+  // TODO: Should the default behavior DO anything?  Also, do these need any
+  // more params?
+  void handleSyclAccessorType(const FieldDecl *, QualType){}
+  void handleSyclSamplerType(const FieldDecl *, QualType){}
+  void handleSyclSpecConstantType(const FieldDecl *, QualType){}
+  void handleStructType(const FieldDecl *, QualType){}
+  void handleReferenceType(const FieldDecl *, QualType){}
+  void handlePointerType(const FieldDecl *, QualType){}
+  void handleArrayType(const FieldDecl *, QualType){}
+  void handleScalarType(const FieldDecl *, QualType){}
+};
+
+// A type to check the valididty of all of the argument types.
+class SYCLKernelFieldChecker : public SyclKernelFieldHandler<SYCLKernelFieldChecker> {
+  bool AllArgsValid = true;
+  DiagnosticsEngine &Diag;
+public:
+  SYCLKernelFieldChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return AllArgsValid; }
+
+  void handleSyclReferenceType(const FieldDecl *FD, QualType ArgTy) {
+    AllArgsValid = false;
+    Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type) << ArgTy;
+  }
+  void handleSyclStructType(const FieldDecl *FD, QualType ArgTy) {
+    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams &&
+        !ArgTy->isStandardLayoutType()) {
+      AllArgsValid = false;
+      Diag.Report(FD->getLocation(), diag::err_sycl_non_std_layout_type)
+          << ArgTy;
+    } else {
+      CXXRecordDecl *RD = ArgTy->getAsCXXRecordDecl();
+      if (!RD->hasTrivialCopyConstructor()) {
+        AllArgsValid = false;
+        Diag.Report(FD->getLocation(),
+                    diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 0 << ArgTy;
+      } else if (!RD->hasTrivialDestructor()) {
+        AllArgsValid = false;
+        Diag.Report(FD->getLocation(),
+                    diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 1 << ArgTy;
+      }
+    }
+  }
+};
+
+// A type to Create and own the FunctionDecl for the kernel.
+class SYCLKernelHeader : public SyclKernelFieldHandler<SYCLKernelHeader> {
```

Comment:
Why this class is needed at all?
What is kernel obj header?

Reply:
I'm not sure yet :)  I was brainstorming based on the current things we do.  The idea was to have this + the error-checking version replace BuildArgTys.  This is more just to build/own the declaration and much of the stuff that happens in ConstructOpenCLKernel.